### PR TITLE
nabla2_...damping refactoring and some stencil tests' xfails cleanup

### PR DESCRIPTION
**Pull request by nfarabullini**
refactoring of `init_nabla2_factor_in_upper_damping_zone` to a gt4py stencil

removal of:

```python
        if isinstance(grid, IconGrid) and grid.limited_area:
            pytest.xfail(
                "Execution domain needs to be restricted or boundary taken into account in stencil."
            )
```
and consequent changes

### CODE REVIEW

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

File: model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_apply_diffusion_to_vn.py [hunk 1]
```diff
@@ -89,15 +90,14 @@ def reference(
                 vn,
             )
 
+        # restriction of execution domain
```
Comment (halungge) [inline model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_apply_diffusion_to_vn.py:93]: Use `helpers.Output`  with slices to restrict comparison domain?

File: model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_calculate_nabla2_for_z.py
Comment (halungge) [inline model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_calculate_nabla2_for_z.py:1]: Use `helpers.Output`  with slices to restrict comparison domain?

**Review by halungge — CHANGES_REQUESTED** [review]
Generally, for the tests why don't you use the `Output` class from  `test_utils.helpers` to restrict the comparison domain? I think that was introduced exactly for that. 

In the diffusion: mostly revise the field allocations that you do.

File: model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_apply_diffusion_to_vn.py
[the remark this answers is not in the document]
Comment (nfarabullini) [reply model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_apply_diffusion_to_vn.py:1]: agreed to leave as is for now

File: model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_apply_diffusion_to_vn.py (hunk 1 above)
Comment (nfarabullini) [reply model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_apply_diffusion_to_vn.py:93]: agreed to leave as is for now

File: model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_calculate_nabla2_for_z.py
Comment (nfarabullini) [reply model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_calculate_nabla2_for_z.py:1]: agreed to leave as is for now

File: model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_apply_diffusion_to_theta_and_exner.py [hunk 2]
```diff
@@ -78,10 +94,9 @@ def reference(
 
     @pytest.fixture
     def input_data(self, grid):
-        if isinstance(grid, IconGrid) and grid.limited_area:
-            pytest.xfail(
-                "Execution domain needs to be restricted or boundary taken into account in stencil."
-            )
+        # error message contained in truly_horizontal_diffusion_nabla_of_theta_over_steep_points_numpy
+        if np.any(grid.connectivities[dims.C2E2CDim] == -1):
```
[the remark this answers is not in the document]
Comment (nfarabullini) [reply model/atmosphere/diffusion/tests/diffusion_stencil_tests/test_apply_diffusion_to_theta_and_exner.py:98]: added TODO

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Review by halungge — APPROVED** [review]
LGTM

**Pull request merged**